extensions/aws/s3/MultipartUploadStateStorage.cpp

```diff
@@ -0,0 +1,127 @@
+
+  int64_t stored_upload_time = 0;
+  core::Property::StringToInt(state_map[state_key + ".upload_time"], stored_upload_time);
+  state.upload_time = Aws::Utils::DateTime(stored_upload_time);
+
+  core::Property::StringToInt(state_map[state_key + ".uploaded_parts"], state.uploaded_parts);
+  core::Property::StringToInt(state_map[state_key + ".uploaded_size"], state.uploaded_size);
+  core::Property::StringToInt(state_map[state_key + ".part_size"], state.part_size);
+  core::Property::StringToInt(state_map[state_key + ".full_size"], state.full_size);
+  state.uploaded_etags = minifi::utils::StringUtils::splitAndTrimRemovingEmpty(state_map[state_key + ".uploaded_etags"], ";");
+  return state;
+}
+
+void MultipartUploadStateStorage::removeKey(const std::string& state_key, std::unordered_map<std::string, std::string>& state_map) {
+  state_map.erase(state_key + ".upload_id");
+  state_map.erase(state_key + ".upload_time");
+  state_map.erase(state_key + ".uploaded_parts");
+  state_map.erase(state_key + ".uploaded_size");
+  state_map.erase(state_key + ".part_size");
+  state_map.erase(state_key + ".full_size");
+  state_map.erase(state_key + ".uploaded_etags");
+}
+
+void MultipartUploadStateStorage::removeState(const std::string& bucket, const std::string& key) {
+  std::unordered_map<std::string, std::string> state_map;
+  if (!state_manager_->get(state_map)) {
+    logger_->log_warn("No previous multipart upload state was associated with this processor.");
+    return;
+  }
+  std::string state_key = bucket + "/" + key;
+  if (!state_map.contains(state_key + ".upload_id")) {
+    logger_->log_warn("Multipart upload state was not found for key '%s'", state_key);
+    return;
+  }
+
+  removeKey(state_key, state_map);
+  state_manager_->set(state_map);
+  state_manager_->commit();
+  state_manager_->persist();
+}
+
+void MultipartUploadStateStorage::removeAgedStates(std::chrono::milliseconds multipart_upload_max_age_threshold) {
+  std::unordered_map<std::string, std::string> state_map;
+  if (!state_manager_->get(state_map)) {
+    logger_->log_warn("No previous multipart upload state was associated with this processor.");
+    return;
+  }
+  auto age_off_time = Aws::Utils::DateTime::Now() - multipart_upload_max_age_threshold;
+
+  std::vector<std::string> keys_to_remove;
+  for (const auto& [property_key, value] : state_map) {
+    if (!minifi::utils::StringUtils::endsWith(property_key, ".upload_time")) {
+      continue;
+    }
+    int64_t stored_upload_time{};
+    if (!core::Property::StringToInt(value, stored_upload_time)) {
+      logger_->log_error("Multipart upload cache key '%s' has invalid value '%s'", property_key, value);
+      continue;
+    }
+    auto upload_time = Aws::Utils::DateTime(stored_upload_time);
+    if (upload_time < age_off_time) {
+      auto state_key_and_property_name = minifi::utils::StringUtils::split(property_key, ".");
+      keys_to_remove.push_back(state_key_and_property_name[0]);
```

Comment:
should we just strip the ".upload_time" suffix instead? as I understand the `property_key` is `<bucket>/<object key>.upload_time` and `object key` is either the `ObjectKey` property set or the `filename` of the flow file, so in the latter case could contain a `.` character for the extension (or hidden file)

Reply:
Good point, updated in 7f27cd5bfdf869971ebab9f706b82405c0e16913